sdk/helper/keysutil/policy.go

```diff
@@ -1355,6 +1357,38 @@ func (p *Policy) Rotate(ctx context.Context, storage logical.Storage, randReader
 		}
 	}()
 
+	if err := p.RotateInMemory(randReader, false); err != nil {
+		return err
+	}
+
+	return p.Persist(ctx, storage)
+}
+
+// RotateInMemory rotates the policy but does not persist it to storage.
+// If restoreOnError is true and the rotation partially fails, the policy
+// state will be restored.
+func (p *Policy) RotateInMemory(randReader io.Reader, restoreOnError bool) (retErr error) {
+	if restoreOnError {
```

Comment:
This feels overcomplicated and prone to unexpected behavior (particularly due to the defer operating on the returned error). What about changing the function such that it runs all of the errorable operations without modifying `p` first. Then at the end, you modify `p` with the appropriate new data.

Reply:
I agree with you wrt how restoring the policy state is handled. This isn't new code being introduced. See that it already exists in [policy.go#L1338-L1356](https://github.com/hashicorp/vault/blob/master/sdk/helper/keysutil/policy.go#L1338-L1356). I've just moved existing code around to support being able to rotate the policy without the final persist.

This code is being used in other places (e.g., transit, transform), so I'm hesitant to modify it and risk introducing bugs.